### ISSUE TITLE
Using `MpSerialExecutor` and `SERIAL_EXEC` still results in multiple processes

### ISSUE THREAD

**Issue by maxzzze** [pytorch/xla#2591]
## 🐛 Bug

From what I understand, when using multiple cores in a TPU with `xpm.spawn` we can utilize `MpSerialExecutor` to run functions only on one process - for example for downloading a dataset only once. While running the following example collab: https://colab.research.google.com/github/pytorch/xla/blob/master/contrib/colab/mnist-training.ipynb#scrollTo=Hx4YVNHametU

I noticed that when I put a print line in the supposedly serialized `get_dataset` function, I still see output printed for each of the 8 cores, indicating what seems to me that each core is still running the function. 

```python
  def get_dataset():
    print("Running get dataset")
    norm = transforms.Normalize((0.1307,), (0.3081,))
    train_dataset = datasets.MNIST(
        FLAGS['datadir'],
        train=True,
        download=True,
        transform=transforms.Compose(
            [transforms.ToTensor(), norm]))
    test_dataset = datasets.MNIST(
        FLAGS['datadir'],
        train=False,
        download=True,
        transform=transforms.Compose(
            [transforms.ToTensor(), norm]))
``` 

> Running get dataset
Downloading http://yann.lecun.com/exdb/mnist/train-images-idx3-ubyte.gz to /tmp/mnist/MNIST/raw/train-images-idx3-ubyte.gz
Extracting /tmp/mnist/MNIST/raw/train-images-idx3-ubyte.gz to /tmp/mnist/MNIST/raw
Downloading http://yann.lecun.com/exdb/mnist/train-labels-idx1-ubyte.gz to /tmp/mnist/MNIST/raw/train-labels-idx1-ubyte.gz
Extracting /tmp/mnist/MNIST/raw/train-labels-idx1-ubyte.gz to /tmp/mnist/MNIST/raw
Downloading http://yann.lecun.com/exdb/mnist/t10k-images-idx3-ubyte.gz to /tmp/mnist/MNIST/raw/t10k-images-idx3-ubyte.gz
Extracting /tmp/mnist/MNIST/raw/t10k-images-idx3-ubyte.gz to /tmp/mnist/MNIST/raw
Downloading http://yann.lecun.com/exdb/mnist/t10k-labels-idx1-ubyte.gz to /tmp/mnist/MNIST/raw/t10k-labels-idx1-ubyte.gz
Extracting /tmp/mnist/MNIST/raw/t10k-labels-idx1-ubyte.gz to /tmp/mnist/MNIST/raw
Processing...
Done!
[xla:0](0) Loss=2.33521 Rate=249.15 GlobalRate=249.14 Time=Fri Oct 30 14:02:00 2020
Running get dataset
[xla:7](0) Loss=2.37399 Rate=819.99 GlobalRate=819.88 Time=Fri Oct 30 14:02:05 2020
Running get dataset
Running get dataset
[xla:4](0) Loss=2.37056 Rate=741.50 GlobalRate=741.42 Time=Fri Oct 30 14:02:07 2020
[xla:2](0) Loss=2.35067 Rate=720.21 GlobalRate=720.13 Time=Fri Oct 30 14:02:08 2020
Running get dataset
[xla:3](0) Loss=2.35181 Rate=678.21 GlobalRate=678.15 Time=Fri Oct 30 14:02:09 2020
Running get dataset
[xla:1](0) Loss=2.33740 Rate=678.90 GlobalRate=677.76 Time=Fri Oct 30 14:02:11 2020
Running get dataset
[xla:5](0) Loss=2.32518 Rate=765.26 GlobalRate=765.15 Time=Fri Oct 30 14:02:13 2020
Running get dataset

## To Reproduce

1. https://colab.research.google.com/github/pytorch/xla/blob/master/contrib/colab/mnist-training.ipynb#scrollTo=Hx4YVNHametU
2. Add print statement to `get_dataset` function in the way described above
3. Run the `xmp.spawn` function

## Expected behavior

Only one print statement because the function is only running on one TPU core.

## Environment

 - Reproducible on XLA backend [CPU/TPU]: TPU
 - torch_xla version: torch-xla-1.6

**Comment by jysohn23** [comment]
`MpSerialExecutor` ensures that the different processes run in serial such that no two processes run the critical section (`get_dataset` fn) concurrently by acquiring the lock created in the global context:

https://github.com/pytorch/xla/blob/ec79c5f4720c493d63ac868f8d3e7804f403d907/torch_xla/distributed/xla_multiprocessing.py#L476

So this behavior is expected, and the reason we do this in this colab notebook example is such that only one process first downloads the dataset onto disk. Subsequent processes will then see that the dataset was downloader and skip the downloading step and just instantiate the dataset.

**Comment by maxzzze** [comment]
> `MpSerialExecutor` ensures that the different processes run in serial such that no two processes run the critical section (`get_dataset` fn) concurrently by acquiring the lock created in the global context:
> 
> https://github.com/pytorch/xla/blob/ec79c5f4720c493d63ac868f8d3e7804f403d907/torch_xla/distributed/xla_multiprocessing.py#L476
> 
> So this behavior is expected, and the reason we do this in this colab notebook example is such that only one process first downloads the dataset onto disk. Subsequent processes will then see that the dataset was downloader and skip the downloading step and just instantiate the dataset.

Understood! Thank you for the detailed explanation @jysohn23 !
